### PR TITLE
Emote Merge

### PR DESCRIPTION
**Pull request by AnatoleAM**
Adding the Emote Merging functionality and a GQL Mutation command

### CODE REVIEW

File: src/server/api/actions/actions.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package actions
```
Comment (TroyDota) [inline src/server/api/actions/actions.go:1]: Why is this here?

File: src/server/api/actions/emote-delete.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+			"status":             datastructure.EmoteStatusDeleted,
+			"last_modified_date": time.Now(),
+		},
+	})
+	if err != nil {
+		log.WithError(err).Error("mongo")
+		return err
+	}
+
+	wg := &sync.WaitGroup{}
```
Comment (TroyDota) [inline src/server/api/actions/emote-delete.go:31]: This doesnt have to be a pointer.

File: src/server/api/actions/emote-merge.go [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+	"github.com/SevenTV/ServerGo/src/mongo"
+	"github.com/SevenTV/ServerGo/src/mongo/datastructure"
+	log "github.com/sirupsen/logrus"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// MergeEmote: Merge an emote into another emote, transferring all its channels and swapping aliases
```
Comment (TroyDota) [inline src/server/api/actions/emote-merge.go:15]: Explain the use case of this.

File: src/server/api/actions/emote-merge.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+	res = mongo.Database.Collection("emotes").FindOne(ctx, bson.M{
+		"_id": opts.NewID,
+	})
+	if err := res.Decode(&newEmote); err != nil {
+		return nil, err
+	}
+
+	switchedChannels := []primitive.ObjectID{}
+	userOps := []mongo.WriteModel{}
+	{
```
Comment (TroyDota) [inline src/server/api/actions/emote-merge.go:50]: Why is this scoped?

File: src/server/api/actions/emote-merge.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+					ch.EmoteAlias[newEmote.ID.Hex()] = oldEmote.Name
+
+				}
+
+				update["emote_alias"] = ch.EmoteAlias
+			}
+
+			// Append a bulk write update operation.
+			// This will be executed later once the merge is confirmed
+			switchedChannels = append(switchedChannels, ch.ID)
```
Comment (TroyDota) [inline src/server/api/actions/emote-merge.go:96]: Can you not do this without a loop?

File: src/server/api/actions/emote-merge.go [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+		}
+		logInfo.Infof("Targeted %d users and updated %d users during merger of Emote(id=%v) into Emote(id=%v)",
+			result.MatchedCount, result.ModifiedCount, oldEmote.ID.Hex(), newEmote.ID.Hex(),
+		)
+	} else {
+		logInfo.Infof("Updated no users during merger of Emote(id=%v) into Emote(id=%v)", oldEmote.ID.Hex(), newEmote.ID.Hex())
+	}
+
+	// Send notifications
+	{
```
Comment (TroyDota) [inline src/server/api/actions/emote-merge.go:126]: Why is this scoped?

File: src/server/api/actions/emote-merge.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+		// Send a notification to the old emote's owner that their emote was merged
+		go Notifications.Create().
```
Comment (TroyDota) [inline src/server/api/actions/emote-merge.go:128]: if the user is the same for multiple notifications, dont add them all?

File: src/server/api/actions/emote-merge.go [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+			AddEmoteMentionPart(newEmote.ID).
+			AddTextMessagePart(" by ").
+			AddUserMentionPart(opts.Actor.ID).
+			AddTextMessagePart(fmt.Sprintf("for the reason \"%v\". %d new channels have been added to your emote and no further action is required.", opts.Reason, len(switchedChannels))).
+			Write(ctx)
+
+	}
+
+	// Now we will delete the old emote
+	Emotes.Delete(ctx, &oldEmote)
```
Comment (TroyDota) [inline src/server/api/actions/emote-merge.go:173]: There is an error here?+

File: src/server/api/v2/gql/resolvers/mutation/emote-merge.go [hunk 9]
```diff
@@ -1 +1,62 @@
...
+		return nil, resolvers.ErrLoginRequired
+	}
+
+	// Check permissions
+	if !usr.HasPermission(datastructure.RolePermissionEmoteEditAll) {
+		return nil, resolvers.ErrAccessDenied
+	}
+
+	// Parse emote IDs
+	var oldID primitive.ObjectID
```
Comment (TroyDota) [inline src/server/api/v2/gql/resolvers/mutation/emote-merge.go:32]: define this as a single var
```go
var (
    oldID primitive.ObjectID
    newID primitive.ObjectID
)
```

Comment (TroyDota) [reply src/server/api/v2/gql/resolvers/mutation/emote-merge.go:32]: and add err. and remove the else checks on the following lines.

**Review by TroyDota — CHANGES_REQUESTED** [review]

File: src/server/api/actions/actions.go (hunk 1 above)
Comment (AnatoleAM) [reply src/server/api/actions/actions.go:1]: I started standardizing common actions on objects so that the code can be re-utilized outside of the GQL environment

File: src/server/api/actions/emote-merge.go (hunk 3 above)
Comment (AnatoleAM) [reply src/server/api/actions/emote-merge.go:15]: https://github.com/SevenTV/ServerGo/pull/45#discussion_r663762426

File: src/server/api/actions/emote-delete.go (hunk 2 above)
Comment (AnatoleAM) [reply src/server/api/actions/emote-delete.go:31]: This is actually your code, I just copy-pasted - this might be something you consider as legacy? May need a full checkup of the codebase to fix other issues like that

File: src/server/api/actions/emote-merge.go (hunk 4 above)
Comment (AnatoleAM) [reply src/server/api/actions/emote-merge.go:50]: Improves readability

Comment (TroyDota) [reply src/server/api/actions/emote-merge.go:50]: WutFace

File: src/server/api/actions/emote-merge.go (hunk 6 above)
Comment (AnatoleAM) [reply src/server/api/actions/emote-merge.go:126]: https://github.com/SevenTV/ServerGo/pull/45#discussion_r663764284

File: src/server/api/actions/emote-merge.go (hunk 4 above)
Comment (TroyDota) [reply src/server/api/actions/emote-merge.go:50]: It looks Weird. WeirdChamp

Comment (AnatoleAM) [reply src/server/api/actions/emote-merge.go:50]: Bruh it looks like straight litty fire fax on a stack wdym

File: src/server/api/actions/emote-merge.go (hunk 3 above)
Comment (TroyDota) [reply src/server/api/actions/emote-merge.go:15]: No, of why someone would want to merge an emote.

Comment (AnatoleAM) [reply src/server/api/actions/emote-merge.go:15]: I mean
This is what this whole pull request is about LULW
it allows mods to merge dupes and whatnot, into the original or a higher quality version

File: src/server/api/actions/emote-merge.go (hunk 5 above)
Comment (AnatoleAM) [reply src/server/api/actions/emote-merge.go:96]: I don't think so, since I need the users' IDs. It's used to target notifications

File: src/server/api/actions/emote-merge.go (hunk 8 above)
Comment (AnatoleAM) [reply src/server/api/actions/emote-merge.go]: Fixed by 7631315c7c91a753eb921072b9b311eede96ebd9

File: src/server/api/v2/gql/resolvers/mutation/emote-merge.go (hunk 9 above)
Comment (AnatoleAM) [reply src/server/api/v2/gql/resolvers/mutation/emote-merge.go]: Fixed in 6f4bd8404b303f9d4740d993498252a5c67ef9b9?

File: src/server/api/actions/emote-merge.go (hunk 7 above)
Comment (AnatoleAM) [reply src/server/api/actions/emote-merge.go:130]: There are 3 notifications being sent to different users / groups of users here

First one is to the owner of the emote being merged: telling them their emote got assrammed and it's gone
Second informs users who had the merged emote enabled that it's been merged into another. (This is fairly unnecessary and can likely be removed in the future, but I figured it'd make for a good stress test)
Third informs the owner of the emote that got merged into that their emote just got X new channels added to it

File: src/server/api/v2/gql/resolvers/mutation/emote-merge.go (hunk 9 above)
Comment (AnatoleAM) [reply src/server/api/v2/gql/resolvers/mutation/emote-merge.go]: Not sure what you meant with remove the else checks, unless you meant I should unscope the err checks? (that uses more lines)

File: src/server/api/actions/emote-merge.go (hunk 3 above)
Comment (TroyDota) [reply src/server/api/actions/emote-merge.go:15]: 👍

File: src/server/api/actions/emote-merge.go [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+					ch.EmoteAlias = map[string]string{}
+				}
+
+				if ok { // The user had an alias for the old emote
+					ch.EmoteAlias[newEmote.ID.Hex()] = a     // Set new alias
+					delete(ch.EmoteAlias, oldEmote.ID.Hex()) // Delete old alias
+
+				} else if oldEmote.Name != newEmote.Name { // Old emote name is different from the emote it is being merged to
+					ch.EmoteAlias[newEmote.ID.Hex()] = oldEmote.Name
+
```
Comment (TroyDota) [inline src/server/api/actions/emote-merge.go:90]: empty line

File: src/server/api/actions/emote-merge.go (hunk 7 above)
Comment (TroyDota) [reply src/server/api/actions/emote-merge.go:129]: Yeah but then in theory 1 person gets 3 notifications :WAYTOODANK: from 1 action?

Comment (AnatoleAM) [reply src/server/api/actions/emote-merge.go:129]: DinkDonk yes, problem? :tf:

File: src/server/api/v2/gql/resolvers/mutation/emote-merge.go (hunk 9 above)
Comment (TroyDota) [reply src/server/api/v2/gql/resolvers/mutation/emote-merge.go]: No, ill do it.

**Review by TroyDota — APPROVED** [review]
👍
